docker/Dockerfile

```diff
@@ -1,51 +1,97 @@
+#   adds OS-level dependencies for *installing* Nautobot, its dependencies, and its development dependencies
+#   installs Poetry
+#   copies Nautobot sourcee files and packaging definition into the image
+#   uses Pip/Poetry to install Nautobot's Python dependencies (but *not* development dependencies or Nautobot itself)
+#   uses Poetry to build a Nautobot wheel (but not yet install it)
+#
+# dependencies-dev (intermediate build target)
+#   installs hadolint for validating Dockerfiles
+#   uses Poetry to additionally install Nautobot's *development* dependencies
+#   adds files and configuration for Nautobot development server to run
+#
+# dev (development environment for Nautobot core)
+#   uses Poetry to install Nautobot as editable
+#     (at runtime the requirement is to mount the live Nautobot source code to /source/; it won't run without that)
+#
+# final-dev (development environment for Nautobot plugins)
+#   uses Pip to install Nautobot as a wheel
+#
+# final (production-ready environment):
+#   removes Poetry
+#   copies all installed Python dependencies from "dependencies"
+#   copies Nautobot wheel from "dependencies" and installs it
+#   creates a self-signed SSL certificate
+#   adds "nautobot" system user
+#   sets up the system to run Nautobot in production with uwsgi
+
+################################ Stage: base (intermediate build target; basis for all other images herein)
+
 FROM python:${PYTHON_VER}-slim as base
 
 ENV PYTHONUNBUFFERED=1 \
     NAUTOBOT_ROOT=/opt/nautobot \
     prometheus_multiproc_dir=/prom_cache
 
-# Install all OS package upgrades and dependencies
+# Install all OS package upgrades and dependencies needed to run Nautobot in production
 # hadolint ignore=DL3005,DL3008,DL3013
 RUN apt-get update && \
     apt-get upgrade -y && \
-    apt-get install --no-install-recommends -y git mime-support curl libxml2 libmariadb3 && \
+    apt-get install --no-install-recommends -y git mime-support curl libxml2 libmariadb3 openssl && \
     apt-get autoremove -y && \
     apt-get clean all && \
     rm -rf /var/lib/apt/lists/* && \
     pip --no-cache-dir install --upgrade pip wheel
 
 HEALTHCHECK --interval=5s --timeout=5s --start-period=5s --retries=1 CMD curl --fail http://localhost:8080/health/ || exit 1
 
-################################ Stage: dev
-# The dev container is used for development purposes but also to build the .whl for the final image
-# nautobot will be installed in editable mode in /source 
-FROM base as dev
+# Generate required dirs for later consumption
+RUN mkdir /opt/nautobot /prom_cache
 
-# Modify the PATH here because otherwise poetry fails 100% of the time. WAT??
-ENV PATH="${PATH}:/root/.local/bin"
+# Common entrypoint for all environments
+COPY docker/docker-entrypoint.sh /docker-entrypoint.sh
+ENTRYPOINT ["/docker-entrypoint.sh"]
+
+################################ Stage: dependencies (intermediate build target)
+
+FROM base as dependencies
+ARG PYUWSGI_VER
```

Comment:
Isn't this redundant since it's defined outside a stage at the top of the dockerfile?

Reply:
Not sure - I was just following the existing pattern in this file for `ARG PYTHON_VER` being declared both at the top of the file and at the build stage that actually uses it.

Reply:
Per my understanding of the ARG behavior, a globally defined ARG doesn't automatically apply within a specific build stage, hence the need to redeclare it within the stages that actually consume it.